plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/UserWrittenCodeTracker.kt

```diff
@@ -0,0 +1,189 @@
+import software.aws.toolkits.jetbrains.services.codewhisperer.language.CodeWhispererProgrammingLanguage
+import software.aws.toolkits.jetbrains.services.codewhisperer.language.programmingLanguage
+import software.aws.toolkits.jetbrains.services.codewhisperer.util.runIfIdcConnectionOrTelemetryEnabled
+import software.aws.toolkits.jetbrains.settings.AwsSettings
+import java.time.Duration
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+
+@Service(Service.Level.PROJECT)
+class UserWrittenCodeTracker(private val project: Project) : Disposable {
+    val userWrittenCodeLineCount = mutableMapOf<CodeWhispererProgrammingLanguage, Long>()
+    val userWrittenCodeCharacterCount = mutableMapOf<CodeWhispererProgrammingLanguage, Long>()
+    private val alarm = AlarmFactory.getInstance().create(Alarm.ThreadToUse.POOLED_THREAD, this)
+
+    private val isShuttingDown = AtomicBoolean(false)
+    val qInvocationCount: AtomicInteger = AtomicInteger(0)
+    private val isQMakingEdits = AtomicBoolean(false)
+    private val isActive: AtomicBoolean = AtomicBoolean(false)
+
+    @Synchronized
+    fun activateTrackerIfNotActive() {
+        // tracker will only be activated if and only if IsTelemetryEnabled = true && isActive = false
+        if (!isTelemetryEnabled() || isActive.get()) return
+        isActive.set(true)
+        // count q service invocations
+        val conn = ApplicationManager.getApplication().messageBus.connect()
+        conn.subscribe(
+            Q_FEATURE_TOPIC,
+            object : QFeatureListener {
+                override fun onEvent(event: QFeatureEvent) {
+                    when (event) {
+                        QFeatureEvent.INVOCATION -> qInvocationCount.getAndIncrement()
+                        QFeatureEvent.STARTS_EDITING -> isQMakingEdits.set(true)
+                        QFeatureEvent.FINISHES_EDITING -> isQMakingEdits.set(false)
+                    }
+                }
+            }
+        )
+        scheduleTracker()
+    }
+
+    fun reset() {
+        userWrittenCodeLineCount.clear()
+        userWrittenCodeCharacterCount.clear()
+        qInvocationCount.set(0)
+        isQMakingEdits.set(false)
+        isActive.set(false)
+        isShuttingDown.set(false)
+    }
+
+    private fun isTelemetryEnabled(): Boolean = AwsSettings.getInstance().isTelemetryEnabled
+
+    private fun scheduleTracker() {
+        if (!alarm.isDisposed && !isShuttingDown.get()) {
+            alarm.addRequest({ flush() }, Duration.ofSeconds(DEFAULT_MODIFICATION_INTERVAL_IN_SECONDS).toMillis())
+        }
+    }
+
+    private fun flush() {
+        try {
+            if (!isTelemetryEnabled() || qInvocationCount.get() <= 0) {
+                return
+            }
+            emitCodeWhispererCodeContribution()
+        } finally {
+            reset()
+            scheduleTracker()
+        }
+    }
+
+    internal fun documentChanged(event: DocumentEvent) {
```

Comment:
maybe assert/annotate that this needs readaction/edt for clarity

Reply:
I added a RequireReadLock of the documentChange function